File: src/com/leetcode/Medium/MinimumRemoveToMakeValidParentheses.java
package com.leetcode.Medium;

import java.util.Stack;

// 1249. Minimum Remove to Make Valid Parentheses
public class MinimumRemoveToMakeValidParentheses {
    public String minRemoveToMakeValid(String s) {
        Stack<Integer> stack = new Stack<>();
        StringBuilder res = new StringBuilder(s);

        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                stack.push(i);
            } else if (s.charAt(i) == ')') {
                if (!stack.isEmpty())
                    stack.pop();
                else
                    res.setCharAt(i, '*');
            }
        }

        while (!stack.isEmpty()) {
            res.setCharAt(stack.pop(), '*');
        }

        return res.toString().replaceAll("\\*", "");
    }
}

File: JavaScript/Medium/MinimumRemoveToMakeValidParentheses.js
/**
 * 1249. Minimum Remove to Make Valid Parentheses
 * @param {string} s
 * @return {string}
 */
const minRemoveToMakeValid = function(s) {
    let stack = [];
    let res = s.split(''); 

    for (let i = 0; i < s.length; i++) {
        if (s.charAt(i) === '(') {
            stack.push(i);
        } else if (s.charAt(i) === ')') {
            if (stack.length > 0)
                stack.pop();
            else 
                res[i] = '*';
        }
    }

    while (stack.length > 0) {
        res[stack.pop()] = '*'; 
    }

    return res.filter(char => char !== '*').join('');
};